tools/database/src/analyse_gas_usage.rs

```diff
@@ -0,0 +1,348 @@
+
+    /// Analyse blocks from the given block height, inclusive
+    #[arg(long)]
+    from_block_height: Option<BlockHeight>,
+
+    /// Analyse blocks up to the given block height, inclusive
+    #[arg(long)]
+    to_block_height: Option<BlockHeight>,
+}
+
+impl AnalyseGasUsageCommand {
+    pub(crate) fn run(&self, home: &PathBuf) -> anyhow::Result<()> {
+        // Create a ChainStore and EpochManager that will be used to read blockchain data.
+        let mut near_config =
+            nearcore::config::load_config(home, near_chain_configs::GenesisValidationMode::Full)
+                .unwrap();
+        let node_storage: NodeStorage = open_storage(&home, &mut near_config).unwrap();
+        let store: Store =
+            node_storage.get_split_store().unwrap_or_else(|| node_storage.get_hot_store());
+        let chain_store = Rc::new(ChainStore::new(
+            store.clone(),
+            near_config.genesis.config.genesis_height,
+            false,
+        ));
+        let epoch_manager =
+            EpochManager::new_from_genesis_config(store, &near_config.genesis.config).unwrap();
+
+        // Create an iterator over the blocks that should be analysed
+        let blocks_iter_opt = crate::block_iterators::make_block_iterator_from_command_args(
+            CommandArgs {
+                last_blocks: self.last_blocks,
+                from_block_height: self.from_block_height,
+                to_block_height: self.to_block_height,
+            },
+            chain_store.clone(),
+        );
+
+        let blocks_iter = match blocks_iter_opt {
+            Some(iter) => iter,
+            None => {
+                println!("No arguments, defaulting to last 100 blocks");
+                Box::new(LastNBlocksIterator::new(100, chain_store.clone()))
+            }
+        };
+
+        // Analyse
+        analyse_gas_usage(blocks_iter, &chain_store, &epoch_manager);
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct GasUsageInShard {
+    pub used_gas_per_account: BTreeMap<AccountId, BigGas>,
+    pub used_gas_total: BigGas,
+}
+
+/// A shard can be split into two halves.
+/// This struct represents the result of splitting a shard at `split_account`.
+#[derive(Debug, Clone)]
+struct ShardSplit {
+    /// Account on which the shard would be split
+    pub split_account: AccountId,
+    /// Gas used by accounts < split_account
+    pub gas_left: BigGas,
+    /// Gas used by accounts >= split_account
+    pub gas_right: BigGas,
+}
+
+impl GasUsageInShard {
+    pub fn new() -> GasUsageInShard {
+        GasUsageInShard { used_gas_per_account: BTreeMap::new(), used_gas_total: 0 }
+    }
+
+    pub fn add_used_gas(&mut self, account: AccountId, used_gas: BigGas) {
+        let old_used_gas: &BigGas = self.used_gas_per_account.get(&account).unwrap_or(&0);
+        let new_used_gas: BigGas = old_used_gas.checked_add(used_gas).unwrap();
+        self.used_gas_per_account.insert(account, new_used_gas);
```

Comment:
Initially I didn't go with the entry API because of `checked_add`, but now I tried it and it's actually pretty nice:
```rust
    pub fn add_used_gas(&mut self, account: AccountId, used_gas: BigGas) {
        let account_gas = self.used_gas_per_account.entry(account).or_insert(0);
        *account_gas = account_gas.checked_add(used_gas).unwrap();
    }
```
:+1: changed to entry api